### PR TITLE
When authenticating with HF OAuth, stay in same tab

### PR DESCRIPTION
**Pull request by Wauplin**
Related to this [slack thread](https://huggingface.slack.com/archives/C061NL7J00Z/p1710245103328939?thread_ts=1710177176.180739&cid=C061NL7J00Z) (internal), https://github.com/huggingface/moon-landing/issues/9054 (internal) and https://github.com/huggingface/moon-landing/pull/9164 (internal).

Thanks to @coyotte508's work in moon-landing , we do not need to open a new tab to login with OAuth in a Gradio Space anymore! This PR simply updates the JS code in `LoginButton` to stay in the same tab during the process. I've also updated the cookie key to invalidate all cookies (necessary). Not a big problem anyway since it will only impact Spaces that will upgrade their Gradio version.

Test Space: https://huggingface.co/spaces/templates/gradio-oauth. To test the full OAuth workflow, don't forget to revoke the permission first in https://huggingface.co/settings/connected-applications.

### CODE REVIEW

**Comment by abidlabs** [comment]
This LGTM. But just a heads up, one issue I noticed in the test space was that I couldn't scroll down to give permissions to specific organizations and then click accept:

image

I think this is due to way that the iframe is set up. Until we fix it, we may want to keep opening in a new tab.

**Review by abidlabs — APPROVED** [review]

**Comment by freddyaboulton** [comment]
@Wauplin do you know if that issue pointed out by @abidlabs will be fixed in the hub?

**Comment by Wauplin** [comment]
@abidlabs @freddyaboulton I also noticed this issue and reported it [on slack](https://huggingface.slack.com/archives/C061NL7J00Z/p1711725869004599) (internal).

File: gradio/components/login_button.py [hunk 1]
```diff
@@ -96,21 +96,11 @@ def _check_login_status(self, request: Request) -> LoginButton:
...
-            window.open(url, '_blank');
-        } else {
-            window.location.assign(url);
-        }
-    } else {
-        url = '/logout' + window.location.search
-        window.location.assign(url);
-    }
+    uri = buttonValue === BUTTON_DEFAULT_VALUE ? '/login/huggingface' : '/logout';
+    window.location.assign(uri + window.location.search);
```
Comment (pngwn) [inline gradio/components/login_button.py:104]: This is where we will need to set the scrolling of the iframe as mention in @coyotte508's PR.

We should be able to add this as the first line of this function:
```js
window.parent?.postMessage({type: "SET_SCROLLING", enabled: true})
```

**Review by pngwn — COMMENTED** [review]
Everything looks good left a comment about the iframe options.

We also need to 'reset' the scrolling when we go back to gradio. This should be straightforward because when we go back to gradio it is a fresh page load, so we can do it somewhere like this:
https://github.com/gradio-app/gradio/blob/main/js/app/index.html#L60-L65

The code to disable scrolling is like so:

```js
window.parent?.postMessage({type: "SET_SCROLLING", enabled: false})
```

**Review by freddyaboulton — APPROVED** [review]
Thanks for addressing the SET_SCROLLING comments @Wauplin ! This looks good to me cc @pngwn

**Comment by pngwn** [comment]
I think it is still not scrolling. lmk if/ when you want me to take a look @Wauplin.

**Comment by Wauplin** [comment]
No it's not working indeed :confused: I posted a message on [slack](https://huggingface.slack.com/archives/C061NL7J00Z/p1712596964662639?thread_ts=1711725869.004599&cid=C061NL7J00Z) (internal), you might have an idea @pngwn?

**Comment by coyotte508** [comment]
It seems on hub's side, I will debug :bow:

**Comment by Wauplin** [comment]
@pete @coyotte508 Thanks for your help! Scrolling now works on this demo Space: https://huggingface.co/spaces/templates/gradio-oauth. I still had to add a 500ms timeout on the redirection to make sure the message is posted (see [here](https://github.com/gradio-app/gradio/pull/7887/files#diff-2b3d7e39d0b46ad26e36b0c01289e571acd7846884de11fb4b1f49d809e0638aR106)). I don't think it downgrades UX but it is not 100% safe on slow connections. Should be ok though? :roll_eyes:

**Comment by coyotte508** [comment]
Posting the message shouldn't be related to the speed of the internet connection, it's child iframe communicating with parent iframe, so I think slow connections will be ok

**Comment by pngwn** [comment]
I'm surprised that such a long timeout is required. I don't think it will cause issues for slow connections. Technically a low machine might have issues but I don't think it is a major concern. A wait of some amount of time is expected though as the message is sent and forgotten about, it doesn't wait for the parent to process it.

File: gradio/components/login_button.py [hunk 2]
```diff
@@ -96,21 +96,16 @@ def _check_login_status(self, request: Request) -> LoginButton:
...
-        }
-    } else {
-        url = '/logout' + window.location.search
-        window.location.assign(url);
-    }
+    uri = buttonValue === BUTTON_DEFAULT_VALUE ? '/login/huggingface' : '/logout';
+    console.log("Setting scrolling to true");
+    window.parent?.postMessage({ type: "SET_SCROLLING", enabled: true }, "*");
+    setTimeout(() => {
+        console.log("Redirecting to " + uri + window.location.search);
```
Comment (pngwn) [inline gradio/components/login_button.py:107]: We should remove these before merging.

**Comment by Wauplin** [comment]
> I'm surprised that such a long timeout is required. 

TBH I only tried with this timeout. If you think it's too long I can update. It's just that it's annoying to test different values (need to push commit, wait for gradio build, then update the Space, wait for rebuild, then test). So if 500ms is fine, I'll keep it.

**Comment by pete** [comment]
> @pete @coyotte508 Thanks for your help!

You are welcome, but I cannot take credit.

**Comment by pngwn** [comment]
> So if 500ms is fine, I'll keep it

it's fine, we'll keep an eye on it.

**Comment by Wauplin** [comment]
Great, then feel free to merge whenever it's ready/approved :)

**Pull request merged**